### ISSUE TITLE
Random panic with caddy plugin

### ISSUE THREAD

**Issue by ueffel** [darkweak/souin#246]
I see the following panic randomly on my raspberry pi. I don't see a pattern yet. Any ideas?

```
panic: Header called after Handler finished
goroutine 79751 [running]:
net/http.(*http2responseWriter).Header(0xbaa2070)
        net/http/h2_bundle.go:6359 +0x80
github.com/darkweak/souin/plugins.(*CustomWriter).Header(...)
        github.com/darkweak/souin@v1.6.18/plugins/base.go:39
github.com/darkweak/souin/plugins/caddy.(*SouinCaddyPlugin).ServeHTTP.func1({0x1d7d938, 0xc4ae130}, 0x69e1f00)
        github.com/darkweak/souin/plugins/caddy@v0.0.0-20220802191641-cb8865f626c1/httpcache.go:129 +0xec
github.com/darkweak/souin/plugins.DefaultSouinPluginCallback.func3()
        github.com/darkweak/souin@v1.6.18/plugins/base.go:187 +0xa0
created by github.com/darkweak/souin/plugins.DefaultSouinPluginCallback
        github.com/darkweak/souin@v1.6.18/plugins/base.go:183 +0x7b0
```

**Comment by darkweak** [comment]
Hello @ueffel I saw that and already worked on a potential fix about the multiple headers sent. Can you try using the hash `3bb68a5508713857f9ecc67308b6f1e621945b28` ?
e.g. `xcaddy build --with github.com/darkweak/souin/plugins/caddy@3bb68a5508713857f9ecc67308b6f1e621945b28 --with github.com/darkweak/souin@3bb68a5508713857f9ecc67308b6f1e621945b28`

**Comment by ueffel** [comment]
I will try it out and keep an eye on it. As I said I have no way to consistently reproduce the error, so a definitive answer if its fixed is hard.

**Comment by ueffel** [comment]
Hmpf it's not even building on my raspberry pi... 

```
# github.com/xujiajun/nutsdb/ds/list
/home/pi/go/pkg/mod/github.com/xujiajun/nutsdb@v0.10.0/ds/list/list.go:267:15: math.MinInt64 (untyped int constant -9223372036854775808) overflows int
```

**Comment by darkweak** [comment]
I can downgrade to v0.9.0 if you want to fix that.

**Comment by darkweak** [comment]
The hash `275879f4a3d51207820be32d2a5b77b975f29e1a` contains the revert bump for nutsdb. Can you try to build with xcaddy using this one? `xcaddy build --with github.com/darkweak/souin/plugins/caddy@275879f4a3d51207820be32d2a5b77b975f29e1a --with github.com/darkweak/souin@275879f4a3d51207820be32d2a5b77b975f29e1a`

**Comment by ueffel** [comment]
xcaddy still tries to build nutsdb v0.10.0. Possible because of this line: <https://github.com/darkweak/souin/blob/275879f4a3d51207820be32d2a5b77b975f29e1a/plugins/caddy/go.mod#L126>

**Comment by darkweak** [comment]
You're right, I'll check the root cause it doesn't use the v0.9.0 only

**Comment by darkweak** [comment]
This one includes the downgrade to `v0.9.0` for caddy: `fdcaf4e74d63d03fbf025487816a181e00809305`

```
xcaddy build --with github.com/darkweak/souin/plugins/caddy@fdcaf4e74d63d03fbf025487816a181e00809305 --with github.com/darkweak/souin@fdcaf4e74d63d03fbf025487816a181e00809305
```

**Comment by ueffel** [comment]
No panics yet, but I noticed that cache-hits are not compressed anymore.

**Comment by darkweak** [comment]
I don't override the `Transfer-Encoding` header so I can't explain that behavior. Does your backend returns a `gzip` HTTP response?

**Comment by ueffel** [comment]
I think the `Content-Encoding` is set from the cache hit and the `encode` module sees that and does nothing.
Meaning `Content-Encoding: gzip` is set, but the response is plain.

**Comment by ueffel** [comment]
```
panic: Header called after Handler finished
goroutine 7844 [running]:
net/http.(*http2responseWriter).Header(0x4ea84a0)
        net/http/h2_bundle.go:6359 +0x80
github.com/darkweak/souin/plugins.(*CustomWriter).Header(...)
        github.com/darkweak/souin@v1.6.19-0.20220828120415-fdcaf4e74d63/plugins/base.go:64
github.com/darkweak/souin/plugins/caddy.(*SouinCaddyPlugin).ServeHTTP.func1({0x1ddf1a0, 0x49fd3e0}, 0x64fe780)
        github.com/darkweak/souin/plugins/caddy@v0.0.0-20220828120415-fdcaf4e74d63/httpcache.go:130 +0x14c
github.com/darkweak/souin/plugins.DefaultSouinPluginCallback.func3({0x1ddf1a0, 0x49fd3e0}, 0x64fe780)
        github.com/darkweak/souin@v1.6.19-0.20220828120415-fdcaf4e74d63/plugins/base.go:231 +0x90
created by github.com/darkweak/souin/plugins.DefaultSouinPluginCallback
        github.com/darkweak/souin@v1.6.19-0.20220828120415-fdcaf4e74d63/plugins/base.go:227 +0x96c
```

Still getting a panic...

**Comment by darkweak** [comment]
I'll rewrite the response streaming implementation then

**Comment by darkweak** [comment]
I reverted everything that could cause the panic in the last commit `698dae89e5eb2c3c726431eeecbeb6f418f3f8b6 `

**Comment by RomanHotsiy** [comment]
We've seen a lot of similar errors recently in our set up with caddy, it happened under the load, maybe relevant.

Do you have some ideas of where we should look in?

We use pretty much default configuration (badger).
One suspicion is that we use rate limit which may respond before the souin? Do you think it may be possible?

Attaching the stacktrace (stacktrace includes `HobMartin/souin`, it's our fork which is already [merged into main](https://github.com/darkweak/souin/pull/526), no other changes).

Thanks in advance 🙏 

```
panic: Header called after Handler finished
--
  | 2024-08-25T18:15:21.606+08:00 | runtime/debug.Stack()
  | 2024-08-25T18:15:21.606+08:00 | runtime/debug/stack.go:24 +0x5e
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.newPanicError({0x21c5300, 0x2d0c540})
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:44 +0x25
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).doCall.func2.1()
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:193 +0x34
  | 2024-08-25T18:15:21.606+08:00 | panic({0x21c5300?, 0x2d0c540?})
  | 2024-08-25T18:15:21.606+08:00 | runtime/panic.go:770 +0x132
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/net/http2.(*responseWriter).Header(0x1d?)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/net@v0.25.0/http2/server.go:2866 +0x6d
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*CustomWriter).Header(0x2724dce?)
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/writer.go:49 +0xb8
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).Store(0xc000002780, 0xc000378e00, 0xc00044e360, 0xc00d492720, {0xc0006500e0, 0x49})
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:219 +0xdfc
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).Upstream.func2()
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:389 +0x545
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).doCall.func2(0xc0008a5b9e, 0xc0090622d0, 0xc000745008?)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:198 +0x64
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).doCall(0x22e3440?, 0xc01abac330?, {0xc0006500e0?, 0x49?}, 0x6cbc8f?)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:200 +0x96
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).Do(0xc0000028c8, {0xc0006500e0, 0x49}, 0xc00ea13ef0)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:113 +0x15a
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).Upstream(0xc000002780, 0xc000378e00, 0xc00044e360, 0xc01cd82740, 0xc00d492720, {0xc0006500e0, 0x49})
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:365 +0x287
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).ServeHTTP.func4(0x7a76c5?, 0xc01a1775f0?)
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:734 +0x36
  | 2024-08-25T18:15:21.606+08:00 | created by github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).ServeHTTP in goroutine 713536
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:733 +0x1490
  | 2024-08-25T18:15:21.606+08:00 | [recovered]
  | 2024-08-25T18:15:21.606+08:00 | panic: Header called after Handler finished
  | 2024-08-25T18:15:21.606+08:00 | runtime/debug.Stack()
  | 2024-08-25T18:15:21.606+08:00 | runtime/debug/stack.go:24 +0x5e
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.newPanicError({0x21c5300, 0x2d0c540})
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:44 +0x25
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).doCall.func2.1()
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:193 +0x34
  | 2024-08-25T18:15:21.606+08:00 | panic({0x21c5300?, 0x2d0c540?})
  | 2024-08-25T18:15:21.606+08:00 | runtime/panic.go:770 +0x132
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/net/http2.(*responseWriter).Header(0x1d?)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/net@v0.25.0/http2/server.go:2866 +0x6d
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*CustomWriter).Header(0x2724dce?)
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/writer.go:49 +0xb8
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).Store(0xc000002780, 0xc000378e00, 0xc00044e360, 0xc00d492720, {0xc0006500e0, 0x49})
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:219 +0xdfc
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).Upstream.func2()
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:389 +0x545
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).doCall.func2(0xc0008a5b9e, 0xc0090622d0, 0xc000745008?)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:198 +0x64
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).doCall(0x22e3440?, 0xc01abac330?, {0xc0006500e0?, 0x49?}, 0x6cbc8f?)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:200 +0x96
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).Do(0xc0000028c8, {0xc0006500e0, 0x49}, 0xc00ea13ef0)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:113 +0x15a
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).Upstream(0xc000002780, 0xc000378e00, 0xc00044e360, 0xc01cd82740, 0xc00d492720, {0xc0006500e0, 0x49})
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:365 +0x287
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).ServeHTTP.func4(0x7a76c5?, 0xc01a1775f0?)
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:734 +0x36
  | 2024-08-25T18:15:21.606+08:00 | created by github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).ServeHTTP in goroutine 713536
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:733 +0x1490
  | 2024-08-25T18:15:21.606+08:00 | goroutine 713569 [running]:
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).Upstream.func1()
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:361 +0xf3
  | 2024-08-25T18:15:21.606+08:00 | panic({0x2331f00?, 0xc007964180?})
  | 2024-08-25T18:15:21.606+08:00 | runtime/panic.go:770 +0x132
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).doCall.func1()
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:170 +0x2b3
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).doCall(0x22e3440?, 0xc01abac330?, {0xc0006500e0?, 0x49?}, 0x6cbc8f?)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:205 +0xb1
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync/singleflight.(*Group).Do(0xc0000028c8, {0xc0006500e0, 0x49}, 0xc00ea13ef0)
  | 2024-08-25T18:15:21.606+08:00 | golang.org/x/sync@v0.7.0/singleflight/singleflight.go:113 +0x15a
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).Upstream(0xc000002780, 0xc000378e00, 0xc00044e360, 0xc01cd82740, 0xc00d492720, {0xc0006500e0, 0x49})
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:365 +0x287
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).ServeHTTP.func4(0x7a76c5?, 0xc01a1775f0?)
  | 2024-08-25T18:15:21.606+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:734 +0x36
  | 2024-08-25T18:15:21.606+08:00 | created by github.com/HobMartin/souin/pkg/middleware.(*SouinBaseHandler).ServeHTTP in goroutine 713536
  | 2024-08-25T18:15:21.607+08:00 | github.com/HobMartin/souin@v0.0.0-20240619095936-9fbc323d41e9/pkg/middleware/middleware.go:733 +0x1490
```

**Comment by darkweak** [comment]
@RomanHotsiy can you try with the latest release? (v1.6.50)

**Comment by darkweak** [comment]
Did you try with souin v1.6.50 instead? There are lot of improvements in the dev repository that are not used in the cache-handler.

**Comment by RomanHotsiy** [comment]
it seems to be fixed on the 1.6.50! ! 🎉 

Thanks for your prompt replies!
